### PR TITLE
output a hint if we get a reset connection w/o error set

### PR DESCRIPTION
**Pull request by weissi**
Motivation:

We have one case in the code where we expect a socket error to be set
when we hit a connection that got the `.reset` event. On Linux that's
expected and therefore we just map this to `ECONNRESET`.
On other platforms however that is unexpected. Now I just found an
[Envoy commit](https://github.com/envoyproxy/envoy/commit/62f8e51bdd2d72cbac739087dabd3f51bdfc5cbd#diff-3315c7b2    91f78a482ebd4cbfc0545362R345)
which suggests this might not work on FreeBSD. This should prompt our
users to let us know if we ever see this error on non-Linux systems.

Modifications:

Add a message that makes it clear this is unexpected and link our
ticket.

Result:

we should learn if this actually happens on macOS.

### CODE REVIEW

**Comment by weissi** [comment]
hit #563 retring

File: Sources/NIO/BaseSocketChannel.swift [hunk 1]
```diff
@@ -921,7 +921,12 @@ class BaseSocketChannel<T: BaseSocket>: SelectableChannel, ChannelCore {
                 } else {
                     // we don't have a socket error, this must be connection reset without an error then
                     // this path should only be executed on Linux (EPOLLHUP, no EPOLLERR)
-                    error = IOError(errnoCode: ECONNRESET, reason: "connection reset (no error set)")
+                    #if os(Linux)
+                    let message: String = "connection reset (no error set)"
+                    #else
+                    let message: String = "BUG IN SwiftNIO (possibly #572), please report! Connection reset (no error set)."
```
Comment (Lukasa) [inline Sources/NIO/BaseSocketChannel.swift]: Should this `precondition` instead?

Comment (weissi) [reply Sources/NIO/BaseSocketChannel.swift]: @Lukasa I was on the fence but it doesn't actually cause any bad behaviour at all. It's "just" potentially not quite the error message in that case.

Comment (Lukasa) [reply Sources/NIO/BaseSocketChannel.swift]: Heh, I suspect that even if this path does execute on Darwin no-one will *ever* report a bug with it. But it doesn't hurt.

**Pull request merged**